kura/org.eclipse.kura.net2.admin/src/main/java/org/eclipse/kura/net2/admin/NMDbusConnector.java

```diff
@@ -0,0 +1,223 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Eurotech and/or its affiliates and others
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *  Eurotech
+ *******************************************************************************/
+package org.eclipse.kura.net2.admin;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.regex.Pattern;
+
+import org.freedesktop.NetworkManager;
+import org.freedesktop.dbus.DBusPath;
+import org.freedesktop.dbus.connections.impl.DBusConnection;
+import org.freedesktop.dbus.exceptions.DBusException;
+import org.freedesktop.dbus.interfaces.Properties;
+import org.freedesktop.dbus.types.UInt32;
+import org.freedesktop.dbus.types.Variant;
+import org.freedesktop.networkmanager.Device;
+import org.freedesktop.networkmanager.Settings;
+import org.freedesktop.networkmanager.settings.Connection;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class NMDbusConnector {
+
+    private static final Logger logger = LoggerFactory.getLogger(NMDbusConnector.class);
+    private static final String NM_BUS_NAME = "org.freedesktop.NetworkManager";
+    private static final String NM_BUS_PATH = "/org/freedesktop/NetworkManager";
+    private static final String NM_SETTINGS_PATH = "/org/freedesktop/NetworkManager/Settings";
+
+    private static NMDbusConnector instance;
+    private DBusConnection dbusConnection;
+    private NetworkManager nm;
+    private Map<String, Object> configurationCache;
+
+    private NMDbusConnector(DBusConnection dbusConnection) throws DBusException {
+        this.dbusConnection = Objects.requireNonNull(dbusConnection);
+        this.nm = this.dbusConnection.getRemoteObject(NM_BUS_NAME, NM_BUS_PATH, NetworkManager.class);
+    }
+
+    public static NMDbusConnector createInstance() throws DBusException {
+        instance = new NMDbusConnector(DBusConnection.getConnection(DBusConnection.DEFAULT_SYSTEM_BUS_ADDRESS));
+        return instance;
+    }
+
+    public static NMDbusConnector createInstance(DBusConnection dbusConnection) throws DBusException {
+        instance = new NMDbusConnector(dbusConnection);
+        return instance;
+    }
+
+    public void closeConnection() {
+        dbusConnection.disconnect();
+    }
+
+    public static NMDbusConnector getInstance() {
+        if (instance == null) {
+            throw new IllegalStateException(
+                    "Instance not created yet. Please use NMDbusConnector.createInstance() first");
+        }
+
+        return instance;
+    }
+
```

Comment:
I think that it is not so clear how this class should be instantiated/created.

Reply:
So yeah... I'm not entirely sold on the current implementation.

The [application](https://github.com/mattdibi/NMTest/tree/feat/version_update) I'm using for testing currently hangs on exit if I don't close the dbus connection. Less than ideal...

I don't think this will remains as-is...

Reply:
Thanks for the early feedback though, really appreciated :D